### DISCUSSION TITLE
adding Dagster Type checks

### DISCUSSION THREAD

**Discussion by thunderbug1** [dagster-io/dagster#20658] [Q&A] [answered]
Hi I am trying  out Dagster and I have the following setup:

```python
def is_valid_ticket_number(_context, obj):
    # Regular expression pattern to match the ticket number format
    pattern = re.compile(r'^[A-Z]+-\d+$')
    return isinstance(obj, str) and pattern.match(obj) is not None

ticket_number_type = DagsterType(
    name="ticket_number",
    description="A ticket number in the format PREFIX-NUMBER, where PREFIX consists of uppercase letters and NUMBER is a sequence of digits",
    type_check_fn=is_valid_ticket_number,
)

@asset(dagster_type=ticket_number_type)
def ticket_numbers(context: AssetExecutionContext):
    dataset_path = os.path.join(os.path.dirname(__file__), "static_files/Ticketnummern_2024-03-20.csv")
    with open(dataset_path, "r") as fd:
        ticket_ids : List[Dict[str, str]] = [row for row in csv.DictReader(fd)]
    context.instance.add_dynamic_partitions(ticket_partitions.name, [row["ticket_number"] for row in ticket_ids])
    return Output(value=ticket_ids, metadata={"samples": MetadataValue.json(ticket_ids[:10])})

```

I get this error:
```
dagster._core.errors.DagsterTypeCheckDidNotPass: Type check failed for step output "result" - expected type "ticket_number". Description: None
```

how would I use the Type correctly?

**Comment by garethbrickman** [comment, marked as answer]
The error message you're encountering indicates that the type check for an output named "result" failed because it did not meet the expected type "ticket_number". 

To address the error, you should ensure that the output of the `ticket_numbers` asset function adheres to the type check defined by `is_valid_ticket_number`. Here are some steps to debug and resolve the issue:

1. Verify the CSV Data:
   Ensure that the CSV file contains valid ticket numbers that match the expected format defined by the regular expression in `is_valid_ticket_number`.

2. Check the Output Type:
   The `ticket_numbers` asset function returns a list of dictionaries, but the custom type check expects a string that matches the ticket number format. You need to modify the asset function to return individual ticket numbers as strings, or adjust the type check to validate the entire list of dictionaries.

3. Update the Type Check Function:
   If you want to keep the output as a list of dictionaries, you need to update the `is_valid_ticket_number` function to iterate over the list and check each ticket number individually.

Here's an example of how you might modify the type check function to handle a list of dictionaries:

```python
def is_valid_ticket_number(_context, obj):
    if not isinstance(obj, list):
        return False
    pattern = re.compile(r'^[A-Z]+-\d+$')
    return all(isinstance(ticket_dict, dict) and 
               isinstance(ticket_dict.get("ticket_number"), str) and 
               pattern.match(ticket_dict["ticket_number"]) is not None 
               for ticket_dict in obj)
```

Remember to update the `ticket_number_type` with the new `type_check_fn` if you make this change.

4. Adjust the Asset Function:
   If you decide to return individual ticket numbers as strings, you need to adjust the `ticket_numbers` asset function to yield each ticket number separately, ensuring that each one passes the type check.

**Reply by thunderbug1** [reply]
That solves the issue thank you!

> asset function to yield each ticket

That is a good hint. I didn't know it can be a generator as well. This makes the validation function a lot cleaner.

Is there a best practice how to name assets? Should it be singular or plural?
When I change the asset to a generator then downstream assets get single tickets as a string as their argument, which makes sense but the argument is still called ticket_numbers which can be confusing as it is just one number.

**Reply by garethbrickman** [reply]
There's the usual software development best practices: 
- Avoid Ambiguity: Ensure that asset names are unique 
- Consistency: Use consistent naming conventions that reflect the asset's purpose, source, and storage location.
- Clarity: Choose names that clearly describe the asset's content and role within your data pipeline.

Dagster has some supplemental features to naming, such as key prefixes, compute kind and doc strings. Can check out some [functional examples like this](https://github.com/dagster-io/hooli-data-eng-pipelines/blob/master/hooli_data_eng/assets/forecasting/__init__.py
) in our [hooli-data-eng-pipelines repo](https://github.com/dagster-io/hooli-data-eng-pipelines).
